kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/PaginatedScanFilesIteratorImpl.java

```diff
@@ -15,52 +15,142 @@
+  private long numScanFilesReturned;
+
+  /**
+   * The name of the last log file that was read. This value corresponds to the one saved in the
+   * page token if present.
+   */
+  private String lastReadLogFileName = null;
+
+  /**
+   * The index of the last row returned from the last log file that was read.
+   *
+   * <p>For example, if the last page contains 3 batches from the same log file, and each batch has
+   * 10 rows, this value will be 29 (since row indices start at 0).
+   *
+   * <p>This value corresponds to the one saved in the page token if present;
+   */
+  private long lastReturnedRowIndex = -1;
+
+  private Optional<FilteredColumnarBatch> currentBatch = Optional.empty();
+  private boolean closed = false;
+
   /**
    * Constructs a paginated iterator over scan files on top of a given filtered scan files iterator
    * and pagination context.
    *
-   * @param filteredScanFilesIter The underlying scan files iterator with data skipping and
+   * @param baseFilteredScanFilesIter The underlying scan files iterator with data skipping and
    *     partition pruning applied. This iterator serves as the source of filtered scan results for
    *     pagination.
    * @param paginationContext The pagination context that carries pagination-related information,
    *     such as the maximum number of files to return in a page.
    */
   public PaginatedScanFilesIteratorImpl(
-      CloseableIterator<FilteredColumnarBatch> filteredScanFilesIter,
+      CloseableIterator<FilteredColumnarBatch> baseFilteredScanFilesIter,
       PaginationContext paginationContext) {
-    this.filteredScanFilesIter = filteredScanFilesIter;
+    this.baseFilteredScanFilesIter = baseFilteredScanFilesIter;
     this.pageSize = paginationContext.getPageSize();
   }
 
   @Override
   public Row getCurrentPageToken() {
-    throw new UnsupportedOperationException("Not implemented");
+    // TODO: change values for data validation here
+    return new PageToken(
+            lastReadLogFileName,
+            lastReturnedRowIndex,
+            Optional.empty() /* sidecar file index */,
+            Meta.KERNEL_VERSION,
+            "fake/table/path" /* table path */,
+            -1 /* table version */,
+            -1 /* predicate hash */,
+            -1 /* log segment hash */)
+        .toRow();
   }
 
   @Override
   public boolean hasNext() {
-    throw new UnsupportedOperationException("Not implemented");
+    if (closed) {
+      throw new IllegalStateException("Can't call `hasNext` on a closed iterator.");
+    }
+    if (!currentBatch.isPresent()) {
+      prepareNext();
+    }
+    return currentBatch.isPresent();
+  }
+
+  private void prepareNext() {
+    if (currentBatch.isPresent()) return;
+    if (numScanFilesReturned >= pageSize) return;
+    if (!baseFilteredScanFilesIter.hasNext()) return;
+
+    FilteredColumnarBatch batch = baseFilteredScanFilesIter.next();
+    checkArgument(batch.getFilePath().isPresent(), "file path doesn't exist!");
```

Comment:
you are comparing `filePath` against `fileName` -- something seems off/wrong here. `/a/b/c/d/e/foo.json` will never be equal to `foo.json`